### ISSUE TITLE
File based QdrantClient does not support multithreading

### ISSUE THREAD

**Issue by pieroit** [qdrant/qdrant-client#367]
Hi all, I am the maintainer of the [Cheshire Cat AI](https://github.com/cheshire-cat-ai/core) (we have Qdrant as default DB).

We currently have this setup:
- using `qdrant_client==1.6.2`
- for tests, in `:memory:` client
- for dev setup, we would like to use file-based Qdrant
- for deploy, the Qdrant container

The issue is about file-based Qdrant, `QdrantClient(path="/something/folder")`.
Since conversations with the cat from different users are handled in different threads, we found no problem in reading the file based DB, but an error occurs when writing:

```
 self.client.upsert(
File "/usr/local/lib/python3.10/site-packages/qdrant_client/qdrant_client.py", line 796, in upsert
  return self._client.upsert(
File "/usr/local/lib/python3.10/site-packages/qdrant_client/local/qdrant_local.py", line 353, in upsert
  collection.upsert(points)
File "/usr/local/lib/python3.10/site-packages/qdrant_client/local/local_collection.py", line 810, in upsert
  self._upsert_point(point)
File "/usr/local/lib/python3.10/site-packages/qdrant_client/local/local_collection.py", line 805, in _upsert_point
  self.storage.persist(point)
File "/usr/local/lib/python3.10/site-packages/qdrant_client/local/persistence.py", line 96, in persist
  cursor = self.storage.cursor()
cheshire_cat_core  | sqlite3.ProgrammingError: SQLite objects created in a thread can only be used in that same thread. The object was created in thread id 139974727231296 and this is thread id 139973477500672.
```

What can we do?

**Comment by pieroit** [comment]
Maybe `check_same_thread=False` when creating the connection?

**Comment by pieroit** [comment]
Just saw there is an [open PR about this](https://github.com/qdrant/qdrant-client/pull/329)

**Comment by joein** [comment]
File based Qdrant is indeed does not support multithreading since it uses SQLite under the hood and SQLite is not threadsafe, it is not safe to write to SQLite from different threads.

I don't think that PR will be merged in its current state, however I think it is possible that we will provide a way to pass arguments to sqlite connection.

**Comment by pieroit** [comment]
> File based Qdrant is indeed does not support multithreading since it uses SQLite under the hood and SQLite is not threadsafe, it is not safe to write to SQLite from different threads.
> 
> I don't think that PR will be merged in its current state, however I think it is possible that we will provide a way to pass arguments to sqlite connection.

File based Qdrant is indeed just for testing/prototyping purpose, but useful to avoid using the whole container (it's one of the main reason we're using Qdrant XD).
Without threading it is not possible to use it in conjunction with any web server! Would be great to have this feat (safe or not)

**Comment by valentimarco** [comment]
On the docs, there is a page dedicated to this topic, as they wrote:
> Use the [SQLITE_THREADSAFE](https://www.sqlite.org/compile.html#threadsafe) compile-time parameter to select the threading mode. If no [SQLITE_THREADSAFE](https://www.sqlite.org/compile.html#threadsafe) compile-time parameter is present, then serialized mode is used. This can be made explicit with [-DSQLITE_THREADSAFE=1](https://www.sqlite.org/compile.html#threadsafe). With [-DSQLITE_THREADSAFE=0](https://www.sqlite.org/compile.html#threadsafe) the threading mode is single-thread. With [-DSQLITE_THREADSAFE=2](https://www.sqlite.org/compile.html#threadsafe) the threading mode is multi-thread.

So is **thread-safe** and they provide also a function (for C API) to check this flag, in python is possible with this query:
```py  
cx.execute("select * from pragma_compile_options where compile_options like 'THREADSAFE=%'").fetchall()
```
Sources:
1. [Using SQLite In Multi-Threaded Applications](https://www.sqlite.org/threadsafe.html)
2. [Test To See If The Library Is Threadsafe](https://www.sqlite.org/c3ref/threadsafe.html)
3. [Is sqlite3.threadsafety the same thing as sqlite3_threadsafe() from the C library?](https://discuss.python.org/t/is-sqlite3-threadsafety-the-same-thing-as-sqlite3-threadsafe-from-the-c-library/11463)

**Comment by joein** [comment]
#369 #376

should be fixed as of qdrant-client 1.6.9
